src/primitives/symmetric/encryption/des/mod.rs

```diff
@@ -0,0 +1,256 @@
+      output[i / 2] |= s_box_value << (4 - (bit_pos % 8));
+    }
+
+    output
+  }
+
+  fn feistel_function(data: &HalfBlock, subkey: &Subkey) -> HalfBlock {
+    let expanded = Self::permutation(data, &E);
+    let mixed = Self::feistel_mix(expanded, subkey);
+    let substituted = Self::feistel_substitution(mixed);
+    Self::permutation(&substituted, &F_P)
+  }
+
+  fn permutation<const N1: usize, const N2: usize>(
+    data: &[u8; N1],
+    lookup_table: &[usize],
+  ) -> [u8; N2] {
+    Self::permute(data, lookup_table)
+      .chunks_exact(8)
+      .map(get_byte_from_bits)
+      .collect::<Vec<u8>>()
+      .try_into()
+      .unwrap_or_else(|v: Vec<u8>| panic!("expected vec of len: {} but got: {}", N2, v.len()))
+  }
+
+  fn mix(state: HalfBlock, subkey: HalfBlock) -> HalfBlock {
+    state
+      .into_iter()
+      .zip(subkey)
+      .map(|(a, b)| a ^ b)
+      .collect::<Vec<u8>>()
+      .try_into()
+      .unwrap_or_else(|v: Vec<u8>| panic!("Expected vec of len: {} but got {}", 4, v.len()))
+  }
+
+  fn des<'a>(data: &Block, subkeys: impl Iterator<Item = &'a Subkey>) -> Block {
+    // initial permutation
+    let ip: [u8; 8] = Self::permutation(data, &IP);
+
+    // split 64-bit block into 2 32-bit blocks
+    let (mut left, mut right) = ([ip[0], ip[1], ip[2], ip[3]], [ip[4], ip[5], ip[6], ip[7]]);
+
+    // 16 rounds
+    for subkey in subkeys.into_iter() {
+      // swap right and left
+      let temp = right;
+      // mix right with round subkey
+      right = Self::mix(left, Self::feistel_function(&right, subkey));
+      left = temp;
+    }
+
+    // final round doesn't swap elements, so we swap back
+    let combined: Block = [right, left]
+      .concat()
+      .try_into()
+      .unwrap_or_else(|v: Vec<u8>| panic!("Expected vec of len: {} but got: {}", 8, v.len()));
+
+    // final permutation
+    Self::permutation(&combined, &FP)
+  }
+
+  /// Encrypt a message of size [`Block`]
+  ///
+  /// ## Example
+  /// ```rust
+  /// use rand::{thread_rng, Rng};
+  /// use ronkathon::primitives::symmetric::encryption::des::DES;
+  /// let mut rng = thread_rng();
+  /// let secret_key = rng.gen();
+  ///
+  /// let des = DES::new(secret_key);
+  ///
+  /// let message = rng.gen();
+  /// let encrypted = des.encrypt(&message);
+  /// ```
+  pub fn encrypt(&self, message: &Block) -> Block { Self::des(message, self.subkeys.iter()) }
+
+  /// Decrypt a ciphertext of size [`Block`]
+  ///
+  /// ## Example
+  /// ```rust
+  /// use rand::{thread_rng, Rng};
+  /// use ronkathon::primitives::symmetric::encryption::des::DES;
+  /// let mut rng = thread_rng();
+  /// let secret_key = rng.gen();
+  ///
+  /// let des = DES::new(secret_key);
+  ///
+  /// let message = rng.gen();
+  /// let encrypted = des.encrypt(&message);
+  /// let decrypted = des.decrypt(&encrypted);
+  /// ```
+  pub fn decrypt(&self, ciphertext: &Block) -> Block {
+    Self::des(ciphertext, self.subkeys.iter().rev())
+  }
```

Comment:
Perhaps these should be part of a trait? 

Would be really cool if we could have a proof that `decrypt(encrypt(msg)) = msg`, but, alas, we are in Rust land not  Lean land.